File: tests/ruby/app/application.rb
# frozen_string_literal: true

require_relative 'services/user_service'
require_relative 'services/notification_service'
require_relative 'services/cache_service'
require_relative 'repositories/user_repository'
require_relative 'models/user'
require_relative '../lib/utils/logger'
require_relative '../lib/modules/authentication_module'
require_relative '../lib/concerns/cacheable'

##
# Main application class demonstrating complex Ruby dependency relationships.
# 
# This application showcases various Ruby patterns including:
# - Module mixins and concerns
# - Service layer architecture
# - Repository pattern
# - Dependency injection
# - Metaprogramming features
# - Block and proc usage
# - Ruby-specific patterns like decorators and validators
class Application
  include AuthenticationModule
  include Cacheable

  attr_reader :user_service, :notification_service, :cache_service, :logger
  
  def initialize(dependencies = {})
    @user_service = dependencies[:user_service] || Services::UserService.new
    @notification_service = dependencies[:notification_service] || Services::NotificationService.new
    @cache_service = dependencies[:cache_service] || Services::CacheService.new
    @logger = dependencies[:logger] || Utils::Logger.new('Application')
    @initialized = false
  end

  ##
  # Initialize the application and all its dependencies.
  def initialize_application
    logger.info('Initializing Crawler Test Ruby Application')
    
    begin
      # Initialize database connections
      setup_database_connections
      
      # Initialize cache service
      cache_service.connect
      logger.info('Cache service connected')
      
      # Initialize services with dependency injection
      user_service.initialize_with_dependencies(
        cache_service: cache_service,
        logger: logger
      )
      
      notification_service.initialize_with_dependencies(
        user_service: user_service,
        cache_service: cache_service,
        logger: logger
      )
      
      @initialized = true
      logger.info('All services initialized successfully')
      
    rescue StandardError => e
      logger.error("Failed to initialize application: #{e.message}")
      logger.error(e.backtrace.join("\n"))
      raise e
    end
  end

  ##
  # Run the main application workflow.
  def run
    raise 'Application not initialized' unless @initialized
    
    logger.info('Starting main application workflow')
    
    begin
      execute_sample_workflow
      logger.info('Main workflow completed successfully')
    rescue StandardError => e
      logger.error("Main workflow failed: #{e.message}")
      raise e
    end
  end

  ##
  # Create a sample user using block syntax and hash parameters.
  def create_sample_user(**params)
    user_data = {
      email: params[:email],
      username: params[:username],
      first_name: params[:first_name],
      last_name: params[:last_name],
      role: params[:role] || :user,
      status: :active,
      email_verified: true
    }
    
    # Use block for validation
    user_service.create_user(user_data) do |user|
      logger.info("Created user: #{user.username} (#{user.id})")
      
      # Send welcome email using proc
      send_welcome_email = proc do |user_instance|
        notification_service.send_welcome_notification(user_instance)
      end
      
      send_welcome_email.call(user)
    end
  end

  private

  ##
  # Execute a sample workflow demonstrating Ruby patterns and service dependencies.
  def execute_sample_workflow
    logger.info('Starting sample workflow execution')
    
    # Create sample users using Ruby's flexible syntax
    users = create_sample_users
    
    # Demonstrate service interactions
    demonstrate_user_operations(users)
    demonstrate_service_interactions(users)
    demonstrate_ruby_patterns(users)
    demonstrate_caching_with_concerns
    
    logger.info('Sample workflow completed')
  end

  ##
  # Create sample users demonstrating Ruby's expressive syntax.
  def create_sample_users
    users = {}
    
    # Use symbols and hash shortcuts
    user_definitions = [
      { email: 'admin@example.com', username: 'admin_user', 
        first_name: 'Admin', last_name: 'User', role: :admin },
      { email: 'moderator@example.com', username: 'mod_user', 
        first_name: 'Moderator', last_name: 'User', role: :moderator },
      { email: 'user@example.com', username: 'regular_user', 
        first_name: 'Regular', last_name: 'User', role: :user }
    ]
    
    # Use each_with_object for functional-style processing
    user_definitions.each_with_object(users) do |definition, hash|
      user = create_sample_user(**definition)
      hash[definition[:role]] = user
    end
  end

  ##
  # Demonstrate user service operations with Ruby idioms.
  def demonstrate_user_operations(users)
    logger.info('Demonstrating user service operations')
    
    # Use safe navigation operator
    active_users = user_service&.get_active_users || []
    logger.info("Found #{active_users.count} active users")
    
    # Demonstrate filtering with blocks
    admin_users = active_users.select(&:admin?)
    moderator_users = active_users.filter { |user| user.role == :moderator }
    
    logger.info("Admin users: #{admin_users.count}, Moderator users: #{moderator_users.count}")
    
    # Use range and case statements
    user_count = active_users.count
    status = case user_count
             when 0
               'No users found'
             when 1..5
               'Few users'
             when 6..20
               'Moderate user base'
             else
               'Large user base'
             end
    
    logger.info("User base status: #{status}")
  end

  ##
  # Demonstrate service interactions using Ruby patterns.
  def demonstrate_service_interactions(users)
    logger.info('Demonstrating service interactions')
    
    # Use parallel assignment
    admin_user, moderator_user, regular_user = users[:admin], users[:moderator], users[:user]
    
    # Send notifications using method chaining
    [regular_user, moderator_user].each do |user|
      notification_service
        .create_notification(user)
        .with_type(:welcome)
        .with_priority(:normal)
        .deliver_later
    end
    
    # Send special admin notification with block
    notification_service.send_notification(admin_user) do |notification|
      notification.type = :admin_welcome
      notification.priority = :high
      notification.template = 'admin_welcome_template'
    end
    
    # Demonstrate user promotion with callbacks
    user_service.promote_user(regular_user.id, to: :moderator) do |promoted_user|
      logger.info("User #{promoted_user.username} promoted to #{promoted_user.role}")
      
      # Send promotion notification
      notification_service.send_promotion_notification(promoted_user)
    end
  end

  ##
  # Demonstrate Ruby-specific patterns and metaprogramming.
  def demonstrate_ruby_patterns(users)
    logger.info('Demonstrating Ruby patterns and metaprogramming')
    
    # Use method_missing delegation pattern
    user_stats = user_service.statistics
    
    # Dynamic method calls
    %w[total_users active_users inactive_users].each do |stat_method|
      value = user_stats.send(stat_method)
      logger.info("#{stat_method}: #{value}")
    end
    
    # Use tap for side effects
    users[:admin].tap do |admin|
      admin.last_login = Time.current
      admin.increment_login_count
      user_service.update_user(admin)
    end
    
    # Demonstrate string interpolation and heredoc
    report = <<~REPORT
      User Report
      ===========
      Total Users: #{users.count}
      Admin Users: #{users.select { |_, user| user.admin? }.count}
      Active Users: #{user_service.get_active_users.count}
      
      Generated at: #{Time.current.strftime('%Y-%m-%d %H:%M:%S')}
    REPORT
    
    logger.info(report)
  end

  ##
  # Demonstrate caching using concerns.
  def demonstrate_caching_with_concerns
    logger.info('Demonstrating caching with concerns')
    
    # Use the Cacheable concern
    cache_key = 'user_statistics'
    
    stats = cached(cache_key, expires_in: 1.hour) do
      logger.info('Calculating fresh user statistics')
      user_service.calculate_detailed_statistics
    end
    
    logger.info("Cached statistics: #{stats}")
    
    # Cache invalidation
    invalidate_cache(cache_key)
    logger.info("Cache invalidated for key: #{cache_key}")
  end

  ##
  # Setup database connections with error handling.
  def setup_database_connections
    # Simulate database connection setup
    logger.info('Setting up database connections')
    
    # Ruby's exception handling with ensure
    begin
      # Database connection logic would go here
      sleep(0.1) # Simulate connection time
      logger.info('Database connections established')
    rescue StandardError => e
      logger.error("Database connection failed: #{e.message}")
      raise e
    ensure
      logger.debug('Database connection attempt completed')
    end
  end

  ##
  # Graceful shutdown of the application.
  def shutdown
    logger.info('Starting application shutdown')
    
    begin
      # Shutdown services in reverse dependency order using safe navigation
      notification_service&.shutdown
      user_service&.shutdown
      cache_service&.disconnect
      
      logger.info('Application shutdown completed successfully')
    rescue StandardError => e
      logger.error("Error during application shutdown: #{e.message}")
    end
  end
end

File: tests/php/src/Application.php
<?php

declare(strict_types=1);

namespace CrawlerTest;

use CrawlerTest\Service\UserService;
use CrawlerTest\Service\NotificationService;
use CrawlerTest\Service\CacheService;
use CrawlerTest\Repository\UserRepository;
use CrawlerTest\Entity\User;
use CrawlerTest\Utils\Logger;
use CrawlerTest\Config\DatabaseConfig;
use Doctrine\ORM\EntityManagerInterface;
use Psr\Log\LoggerInterface;

/**
 * Main application class demonstrating complex PHP dependency relationships.
 * 
 * This application showcases various patterns including:
 * - Dependency injection
 * - Service layer architecture  
 * - Repository pattern with Doctrine ORM
 * - Interface-based design
 * - Namespace organization
 * - Trait usage for shared functionality
 */
class Application
{
    private UserService $userService;
    private NotificationService $notificationService;
    private CacheService $cacheService;
    private LoggerInterface $logger;
    private EntityManagerInterface $entityManager;

    public function __construct(
        UserService $userService,
        NotificationService $notificationService,
        CacheService $cacheService,
        LoggerInterface $logger,
        EntityManagerInterface $entityManager
    ) {
        $this->userService = $userService;
        $this->notificationService = $notificationService;
        $this->cacheService = $cacheService;
        $this->logger = $logger;
        $this->entityManager = $entityManager;
    }

    /**
     * Initialize the application and all its dependencies.
     */
    public function initialize(): void
    {
        $this->logger->info('Initializing Crawler Test PHP Application');

        try {
            // Initialize database connection
            $this->entityManager->getConnection()->connect();
            $this->logger->info('Database connection established');

            // Initialize cache service
            $this->cacheService->connect();
            $this->logger->info('Cache service initialized');

            // Initialize services
            $this->userService->initialize();
            $this->notificationService->initialize();

            $this->logger->info('All services initialized successfully');

        } catch (\Exception $e) {
            $this->logger->error('Failed to initialize application', [
                'error' => $e->getMessage(),
                'trace' => $e->getTraceAsString()
            ]);
            throw $e;
        }
    }

    /**
     * Run the main application workflow.
     */
    public function run(): void
    {
        $this->logger->info('Starting main application workflow');

        try {
            $this->executeSampleWorkflow();
            $this->logger->info('Main workflow completed successfully');

        } catch (\Exception $e) {
            $this->logger->error('Main workflow failed', [
                'error' => $e->getMessage()
            ]);
            throw $e;
        }
    }

    /**
     * Execute a sample workflow demonstrating service dependencies.
     */
    private function executeSampleWorkflow(): void
    {
        $this->logger->info('Starting sample workflow execution');

        // Create sample users with different roles
        $adminUser = $this->createSampleUser(
            'admin@example.com',
            'admin_user',
            'Admin',
            'User',
            User::ROLE_ADMIN
        );

        $moderatorUser = $this->createSampleUser(
            'moderator@example.com',
            'mod_user',
            'Moderator',
            'User',
            User::ROLE_MODERATOR
        );

        $regularUser = $this->createSampleUser(
            'user@example.com',
            'regular_user',
            'Regular',
            'User',
            User::ROLE_USER
        );

        $this->logger->info('Created sample users', [
            'admin_id' => $adminUser->getId(),
            'moderator_id' => $moderatorUser->getId(),
            'user_id' => $regularUser->getId()
        ]);

        // Demonstrate service interactions
        $this->demonstrateUserOperations();
        $this->demonstrateServiceInteractions($adminUser, $moderatorUser, $regularUser);
        $this->demonstrateCachingOperations();

        $this->logger->info('Sample workflow completed');
    }

    /**
     * Create a sample user with specified details.
     */
    private function createSampleUser(
        string $email,
        string $username,
        string $firstName,
        string $lastName,
        string $role
    ): User {
        $userData = [
            'email' => $email,
            'username' => $username,
            'firstName' => $firstName,
            'lastName' => $lastName,
            'role' => $role,
            'status' => User::STATUS_ACTIVE,
            'emailVerified' => true
        ];

        return $this->userService->createUser($userData);
    }

    /**
     * Demonstrate various user service operations.
     */
    private function demonstrateUserOperations(): void
    {
        $this->logger->info('Demonstrating user service operations');

        // Get active users
        $activeUsers = $this->userService->getActiveUsers();
        $this->logger->info('Found active users', ['count' => count($activeUsers)]);

        // Get users by role
        $adminUsers = $this->userService->getUsersByRole(User::ROLE_ADMIN);
        $this->logger->info('Found admin users', ['count' => count($adminUsers)]);

        // Demonstrate user search
        $searchResults = $this->userService->searchUsers('admin');
        $this->logger->info('User search results', ['count' => count($searchResults)]);

        // Get user statistics
        $stats = $this->userService->getUserStatistics();
        $this->logger->info('User statistics', $stats);
    }

    /**
     * Demonstrate interactions between different services.
     */
    private function demonstrateServiceInteractions(User $adminUser, User $moderatorUser, User $regularUser): void
    {
        $this->logger->info('Demonstrating service interactions');

        // Send welcome notifications
        $this->notificationService->sendWelcomeNotification($regularUser);
        $this->notificationService->sendWelcomeNotification($moderatorUser);
        $this->notificationService->sendAdminWelcomeNotification($adminUser);

        // Demonstrate user role promotion
        $this->userService->promoteUser($regularUser->getId(), User::ROLE_MODERATOR);
        $this->logger->info('Promoted user to moderator', [
            'user_id' => $regularUser->getId()
        ]);

        // Send bulk notifications
        $userIds = [
            $adminUser->getId(),
            $moderatorUser->getId(),
            $regularUser->getId()
        ];

        $this->userService->sendBulkNotification(
            $userIds,
            'System maintenance scheduled for tonight'
        );

        $this->logger->info('Sent bulk notification', [
            'recipient_count' => count($userIds)
        ]);
    }

    /**
     * Demonstrate caching operations and patterns.
     */
    private function demonstrateCachingOperations(): void
    {
        $this->logger->info('Demonstrating caching operations');

        // Cache user data
        $users = $this->userService->getActiveUsers();
        $cacheKey = 'active_users_list';
        
        $this->cacheService->set($cacheKey, $users, 3600); // Cache for 1 hour
        $this->logger->info('Cached active users list');

        // Retrieve from cache
        $cachedUsers = $this->cacheService->get($cacheKey);
        if ($cachedUsers !== null) {
            $this->logger->info('Retrieved users from cache', [
                'count' => count($cachedUsers)
            ]);
        }

        // Demonstrate cache invalidation
        $this->cacheService->delete($cacheKey);
        $this->logger->info('Invalidated cache key', ['key' => $cacheKey]);

        // Demonstrate cache tags
        $this->cacheService->setWithTags(
            'user_stats',
            $this->userService->getUserStatistics(),
            ['users', 'statistics'],
            7200 // 2 hours
        );

        // Invalidate by tags
        $this->cacheService->invalidateByTags(['users']);
        $this->logger->info('Invalidated cache by tags', ['tags' => ['users']]);
    }

    /**
     * Graceful shutdown of the application.
     */
    public function shutdown(): void
    {
        $this->logger->info('Starting application shutdown');

        try {
            // Shutdown services in reverse dependency order
            $this->notificationService->shutdown();
            $this->userService->shutdown();
            $this->cacheService->disconnect();
            $this->entityManager->getConnection()->close();

            $this->logger->info('Application shutdown completed successfully');

        } catch (\Exception $e) {
            $this->logger->error('Error during application shutdown', [
                'error' => $e->getMessage()
            ]);
        }
    }
}